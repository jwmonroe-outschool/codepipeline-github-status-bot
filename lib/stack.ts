import * as path from "path";

import { Arn, Construct, Stack, StackProps } from "@aws-cdk/core";
import { Function, Code, Runtime } from "@aws-cdk/aws-lambda";
import { Rule } from "@aws-cdk/aws-events";
import { LambdaFunction } from "@aws-cdk/aws-events-targets";
import { Secret } from "@aws-cdk/aws-secretsmanager";
import { PolicyStatement } from "@aws-cdk/aws-iam";

export class CodepipelineGithubStatusBotStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const secretName = "outschool/cdk/githubAccessToken";

    const lambda = new Function(this, "Function", {
      runtime: Runtime.NODEJS_12_X,
      code: Code.fromAsset(path.join(__dirname, "../src/")),
      handler: "index.handler",
      environment: {
        GITHUB_ACCESS_TOKEN_SECRET_NAME: secretName,
      },
    });

    if (!lambda.role) {
      throw new Error("Lambda has no execution role, this should not happen");
    }

    lambda.role.addToPolicy(
      new PolicyStatement({
        actions: ["codepipeline:GetPipelineExecution"],
        resources: ["*"],
      })
    );

    const secret = Secret.fromSecretArn(
      this,
      "Secret",
      Arn.format(
        {
          service: "secretsmanager",
          resource: "secret",
          sep: ":",
          resourceName: `${secretName}*`,
        },
        this
      )
    );

    secret.grantRead(lambda.role);

    new Rule(this, "Rule", {
      eventPattern: {
        source: ["aws.codepipeline"],
        detailType: ["CodePipeline Stage Execution State Change"],
      },
      targets: [new LambdaFunction(lambda)],
    });
  }
}
