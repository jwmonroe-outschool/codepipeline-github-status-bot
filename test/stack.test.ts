import { SynthUtils } from "@aws-cdk/assert";
import { App } from "@aws-cdk/core";
import { CodepipelineGithubStatusBotStack } from "../lib/stack";

test("it should match the latest snapshot", () => {
  const app = new App();
  const stack = new CodepipelineGithubStatusBotStack(app, "MyTestStack");
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
