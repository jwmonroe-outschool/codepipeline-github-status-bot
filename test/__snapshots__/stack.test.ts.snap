// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should match the latest snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersb31e6fee44b2a91cfc60133388e1688f50fc6d53727665a48e2fe7ce51656d76ArtifactHashA958FFEC": Object {
      "Description": "Artifact hash for asset \\"b31e6fee44b2a91cfc60133388e1688f50fc6d53727665a48e2fe7ce51656d76\\"",
      "Type": "String",
    },
    "AssetParametersb31e6fee44b2a91cfc60133388e1688f50fc6d53727665a48e2fe7ce51656d76S3BucketEB722ADC": Object {
      "Description": "S3 bucket for asset \\"b31e6fee44b2a91cfc60133388e1688f50fc6d53727665a48e2fe7ce51656d76\\"",
      "Type": "String",
    },
    "AssetParametersb31e6fee44b2a91cfc60133388e1688f50fc6d53727665a48e2fe7ce51656d76S3VersionKey3382A362": Object {
      "Description": "S3 key for asset version \\"b31e6fee44b2a91cfc60133388e1688f50fc6d53727665a48e2fe7ce51656d76\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "Function76856677": Object {
      "DependsOn": Array [
        "FunctionServiceRoleDefaultPolicy2F49994A",
        "FunctionServiceRole675BB04A",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersb31e6fee44b2a91cfc60133388e1688f50fc6d53727665a48e2fe7ce51656d76S3BucketEB722ADC",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb31e6fee44b2a91cfc60133388e1688f50fc6d53727665a48e2fe7ce51656d76S3VersionKey3382A362",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb31e6fee44b2a91cfc60133388e1688f50fc6d53727665a48e2fe7ce51656d76S3VersionKey3382A362",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "GITHUB_ACCESS_TOKEN_SECRET_NAME": "outschool/cdk/githubAccessToken",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "FunctionServiceRole675BB04A",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "FunctionAllowEventRuleMyTestStackRule0FDD01B9501EA31F": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "Function76856677",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "Rule4C995B7F",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "FunctionServiceRole675BB04A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "FunctionServiceRoleDefaultPolicy2F49994A": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "codepipeline:GetPipelineExecution",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "secretsmanager:GetSecretValue",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":secretsmanager:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":secret:outschool/cdk/githubAccessToken*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "FunctionServiceRoleDefaultPolicy2F49994A",
        "Roles": Array [
          Object {
            "Ref": "FunctionServiceRole675BB04A",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "Rule4C995B7F": Object {
      "Properties": Object {
        "EventPattern": Object {
          "detail-type": Array [
            "CodePipeline Stage Execution State Change",
          ],
          "source": Array [
            "aws.codepipeline",
          ],
        },
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "Function76856677",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
  },
}
`;
